### PR TITLE
[AArch64][GlobalISel] Improve codegen for G_VECREDUCE_{SMIN,SMAX,UMIN,UMAX} for odd-sized vectors

### PR DESCRIPTION
**Pull request by dc03-work**
i8 vectors do not have their sizes changed as I noticed regressions in some tests when that was done.

This patch also adds support for most G_VECREDUCE_* operations to moreElementsVector in LegalizerHelper.cpp.

The code for getting the "neutral" element is taken almost exactly as it is in SelectionDAG, with the exception that support for G_VECREDUCE_{FMAXIMUM,FMINIMUM} was not added.

The code for SelectionDAG is located at SelectionDAG::getNeutralELement().

### CODE REVIEW

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 1]
```diff
@@ -1074,6 +1074,13 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
                  {s16, v8s16},
                  {s32, v2s32},
                  {s32, v4s32}})
+      .moreElementsIf(
+          [=](const LegalityQuery &Query) {
+            return Query.Types[1].isVector() &&
+                   Query.Types[1].getElementType() != s8 &&
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:1080]: The i8 issues look like they might be coming from the calling convention of v3i8 types. They might look better of the inputs were vectors. They might not be quite correct at the moment if they were enabled though. I wouldn't be against just using moreElementsToNextPow2 in the long run, if it doesn't need a more precise "cost model" than that.

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 2]
```diff
@@ -5420,6 +5456,35 @@ LegalizerHelper::moreElementsVector(MachineInstr &MI, unsigned TypeIdx,
...
+  case TargetOpcode::G_VECREDUCE_AND:
+  case TargetOpcode::G_VECREDUCE_OR:
+  case TargetOpcode::G_VECREDUCE_XOR:
+  case TargetOpcode::G_VECREDUCE_SMAX:
+  case TargetOpcode::G_VECREDUCE_SMIN:
+  case TargetOpcode::G_VECREDUCE_UMAX:
+  case TargetOpcode::G_VECREDUCE_UMIN: {
+    LLT OrigTy = MRI.getType(MI.getOperand(1).getReg());
+    MachineOperand &MO = MI.getOperand(1);
+    auto NewVec = MIRBuilder.buildPadVectorWithUndefElements(MoreTy, MO);
```
[the remark this answers is not in the document]
Comment (dc03-work) [reply llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:5472]: Using G_BUILD_VECTOR seems to cause quite a few regressions: https://gist.github.com/dc03-work/66c179f4d380cfac3ed7b36525a583ef.

[the remark this answers is not in the document]
Comment (davemgreen) [reply llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:5473]: It sounds like there is a combine to clear up the undef version, but not the direct one.

**Review by davemgreen — APPROVED** [review]
In that case this LGTM. Thanks for the patch!

**Pull request merged**